lci_strategic_plugin/src/lci_strategic_plugin.cpp

```diff
@@ -608,37 +608,66 @@ void LCIStrategicPlugin::planWhenAPPROACHING(const cav_srvs::PlanManeuversReques
                                           + ros::Duration(EPSILON); //0.01sec more buffer since green_light algorithm's timestamp picks the previous signal - Vehicle Estimation
-    is_entry_time_within_future_events = true; 
+    is_entry_time_within_green_or_tdb = true; 
   }
   else if(config_.enable_carma_streets_connection ==true && scheduled_enter_time_ != 0 ) // UC3
   {
     nearest_green_entry_time = ros::Time(std::max(earliest_entry_time.toSec(), (scheduled_enter_time_)/1000.0)) + ros::Duration(EPSILON); //Carma Street 
     
     // check if scheduled_enter_time_ is inside the available states interval
     size_t i = 0;
+    
+
     for (auto pair : traffic_light->recorded_time_stamps)
     {
-      if (pair.second == lanelet::CarmaTrafficSignalState::PROTECTED_MOVEMENT_ALLOWED && 
-        lanelet::time::timeFromSec(nearest_green_entry_time.toSec()) < pair.first ) 
+      if (lanelet::time::timeFromSec(nearest_green_entry_time.toSec()) < pair.first)
       {
-        ROS_DEBUG_STREAM("ET is inside the available future events! where starting time: " << std::to_string(lanelet::time::toSec(traffic_light->recorded_start_time_stamps[i])) 
-          << ", ending time of that green signal is: " << std::to_string(lanelet::time::toSec(pair.first)));
-        is_entry_time_within_future_events = true;
+        if (pair.second == lanelet::CarmaTrafficSignalState::PROTECTED_MOVEMENT_ALLOWED) 
+        {
+          ROS_DEBUG_STREAM("ET is inside the GREEN phase! where starting time: " << std::to_string(lanelet::time::toSec(traffic_light->recorded_start_time_stamps[i])) 
+            << ", ending time of that green signal is: " << std::to_string(lanelet::time::toSec(pair.first)));
+          is_entry_time_within_green_or_tdb = true;
+        }
+        else
+        {
+          ROS_ERROR_STREAM("Vehicle should plan cruise and stop as ET is inside the RED or YELLOW phase! where starting time: " << std::to_string(lanelet::time::toSec(traffic_light->recorded_start_time_stamps[i])) 
+            << ", ending time of that green signal is: " << std::to_string(lanelet::time::toSec(pair.first)));
+          is_entry_time_within_green_or_tdb = false;
+        }
+
+        in_tdb = false;
         break;
       }
       i++;
     }
+
+    if (in_tdb)
+    {
+      ROS_DEBUG_STREAM("ET is inside TDB phase! where starting time: " << std::to_string(lanelet::time::toSec(traffic_light->recorded_time_stamps.back().first)));
+      is_entry_time_within_green_or_tdb = true;
+    }
+
   }
+
+ 
   
   ROS_DEBUG_STREAM("nearest_green_entry_time: " << std::to_string(nearest_green_entry_time.toSec()) << ", with : " << nearest_green_entry_time - current_state.stamp  << " seconds left at: " << std::to_string(current_state.stamp.toSec()));
   
-  if (!nearest_green_entry_time_cached_ && is_entry_time_within_future_events) 
+  if (nearest_green_entry_time_cached_) 
```

Comment:
This check might change the nearest_green_entry_time to nearest_green_entry_time_cached_. I agree that in this case, nearest_green_entry_time_cached_ is most likely within a green. But still, I feel it is better to move this check to line 623 where we calculate nearest_green_entry_time. Then we can check the if the final nearest_green_entry_time is within green (tbd) or not.

Reply:
- I think you concern is when last ET was in TDB (cached), but new calculated ET is closer and in green, lci might overwrite the new valid green ET with old TDB ET. If that's the case, lci resets nearest_green_entry_time_cached_ everytime it gets a new ET (it checks raw scheduled_entry_time values with old one) from the street, so this cached_ is only valid for the given ET. So it should work correctly. Let me know if you had other case in mind,

- However, the main reason I left it here is because of UC2 logic though. Taking it out to somewhere else might break UC2 or having to have similar code in two places.

Reply:
https://github.com/usdot-fhwa-stol/carma-platform/blob/feature/uc3_et_logic/lci_strategic_plugin/src/lci_strategic_plugin.cpp#L999

this is where it resets cached

Reply:
If every time we receive a new ET that is different from the previous ET, we set the cached one to null, then I guess we will be fine. Because in this case, ET will be either equal to the previous one or the new one which is already checked over spat. The only other case I can think of is if the ET update comes later than the SPaT update. In that case, it is possible that we have an old ET with a new SPaT and the checks over spat will have different results. But I am not sure how much we need be worry about this case.

Reply:
From logs, I can confirm that new SPAT is received every 0.1s while the ET is logged every 1s, so I am assuming that chance is very low.

Reply:
Ok. This will leave 1 sec for this possible issue. We can leave it for now, but again, let keep it in mind in case we see issues in testing. We might want to save these conversations somewhere.

Reply:
This is a very valid reason although unlikely. So we decided to keep it the same just to keep it same semantics as UC2 to not break anything. But will come back here if anything weird is noticed in the logs.